### PR TITLE
Apsides

### PR DESCRIPTION
**Pull request by eggrobin**
Flight plan apsides in map view, cleanup and splitting of the C#.
The C++ interface is temporary and obviously broken, to be rewritten immediately as we remove `LineAndIterator`.
No prediction apsides yet, I need to figure out a way to make them look distinct from the flight plan apsides somehow, so that's still part of #953.

### CODE REVIEW

File: ksp_plugin/interface.cpp [hunk 1]
```diff
@@ -626,6 +626,33 @@ LineAndIterator* principia__RenderedPrediction(
...
+                                          char const* const vessel_guid,
+                                          int const celestial_index,
+                                          XYZ const sun_world_position,
+                                          LineAndIterator** const apoapsides,
+                                          LineAndIterator** const periapsides) {
+  journal::Method<journal::RenderedPredictionApsides> m(
+      {plugin, vessel_guid, celestial_index, sun_world_position},
+      {apoapsides, periapsides});
+  CHECK_NOTNULL(plugin);
+  auto const& prediction = plugin->GetVessel(vessel_guid)->prediction();
```
Comment (pleroy) [inline ksp_plugin/interface.cpp]: CHECK_NOTNULL around GetVessel.

File: ksp_plugin/interface.hpp [hunk 2]
```diff
@@ -53,6 +54,11 @@ R3Element<double> ToR3Element(XYZ const& xyz);
 WXYZ ToWXYZ(Quaternion const& quaternion);
 XYZ ToXYZ(R3Element<double> const& r3_element);
 
+not_null<std::unique_ptr<LineAndIterator>> RenderApsides(
```
Comment (pleroy) [inline ksp_plugin/interface.hpp]: We should not have random stuff in interface.hpp.  This function should be in the unnamed namespace of interface.cpp.

File: ksp_plugin/interface_flight_plan.cpp [hunk 3]
```diff
@@ -263,6 +263,34 @@ void principia__FlightPlanRemoveLast(Plugin const* const plugin,
...
+                                          XYZ const sun_world_position,
+                                          LineAndIterator** const apoapsides,
+                                          LineAndIterator** const periapsides) {
+  journal::Method<journal::FlightPlanRenderedApsides> m(
+      {plugin, vessel_guid, celestial_index, sun_world_position},
+      {apoapsides, periapsides});
+  DiscreteTrajectory<Barycentric>::Iterator begin;
+  DiscreteTrajectory<Barycentric>::Iterator end;
+  CHECK_NOTNULL(plugin);
+  plugin->GetVessel(vessel_guid)->flight_plan().GetAllSegments(&begin, &end);
```
Comment (pleroy) [inline ksp_plugin/interface_flight_plan.cpp]: Use GetFlightPlan.

File: ksp_plugin/plugin.cpp [hunk 4]
```diff
@@ -66,9 +66,14 @@ Permutation<World, AliceWorld> const kWorldLookingGlass(
...
 
-Time const kStep = 45 * Minute;
 Length const kFittingTolerance = 1 * Milli(Metre);
 
+Ephemeris<Barycentric>::FixedStepParameters DefaultEphemerisParameters() {
+  return Ephemeris<Barycentric>::FixedStepParameters(
+             McLachlanAtela1992Order5Optimal<Position<Barycentric>>(),
+             /*step=*/45 * Minute);
+}
+
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Unclear why this diff shows up here.  Is there anything significant?

File: ksp_plugin/plugin.cpp [hunk 5]
```diff
@@ -413,6 +418,19 @@ RenderedTrajectory<World> Plugin::RenderedTrajectoryFromIterators(
   return result;
 }
 
+void Plugin::ComputeApsides(
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: I'd change this function to do ComputeApsides+RenderApsides+RenderApsides.

File: ksp_plugin/interface_flight_plan.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/interface_flight_plan.cpp]: But why is that even a thing?

File: ksp_plugin/plugin.cpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: But we tend to do the rendering in the interface now...

File: ksp_plugin_adapter/map_node_pool.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+         KSP.UI.Screens.Mapview.MapNode.IconData icon) => {
+          CelestialBody celestial = properties_[node].celestial;
+          UnityEngine.Color colour =
+              celestial.orbit == null
+                  ? XKCDColors.SunshineYellow
+                  : celestial.orbitDriver.Renderer.nodeColor;
+          colour.a = 1;
+          icon.visible = true;
+          if (properties_[node].celestial.GetAltitude(
+              properties_[node].world_position) < 0) {
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: Bad indent.

File: ksp_plugin_adapter/map_node_pool.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            // Make sure we see impacts.
+            colour = XKCDColors.Orange;
+          }
+          icon.color = colour;
+        };
+    new_node.OnUpdateType +=
+        (KSP.UI.Screens.Mapview.MapNode node,
+         KSP.UI.Screens.Mapview.MapNode.TypeData type) => {
+          if (properties_[node].celestial.GetAltitude(
+              properties_[node].world_position) < 0) {
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: Same.

File: ksp_plugin/interface_flight_plan.cpp (hunk 3 above)
Comment (pleroy) [inline ksp_plugin/interface_flight_plan.cpp]: To do the checks that you are missing here.

File: ksp_plugin/interface.cpp (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin/interface.cpp]: `CHECK_NOTNULL` of a `not_null`?

File: ksp_plugin/interface_flight_plan.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/interface_flight_plan.cpp]: No, to log the GUID if they fail, the same checks are performed here.

File: ksp_plugin/interface.hpp (hunk 2 above)
Comment (eggrobin) [inline ksp_plugin/interface.hpp]: Done.

File: ksp_plugin/interface_flight_plan.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/interface_flight_plan.cpp]: Done.

File: ksp_plugin_adapter/map_node_pool.cs (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin_adapter/map_node_pool.cs]: Done.

File: ksp_plugin_adapter/map_node_pool.cs (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin_adapter/map_node_pool.cs]: Done.

File: ksp_plugin/plugin.cpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/plugin.cpp]: Done.

**Pull request merged**